### PR TITLE
[AIRFLOW-489] Add API Framework

### PR DESCRIPTION
**Pull request by bolkedebruin**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- AIRFLOW-489

Per Apache guidelines you need to create a [Jira issue](https://issues.apache.org/jira/browse/AIRFLOW/).

Testing Done:
- to be added

Reminders for contributors (REQUIRED!):
- Your PR's title must reference an issue on 
  [Airflow's JIRA](https://issues.apache.org/jira/browse/AIRFLOW/). 
  For example, a PR called "[AIRFLOW-1] My Amazing PR" would close JIRA 
  issue #1. Please open a new issue if required!
- Please squash your commits when possible and follow the [How to write a good git commit message](http://chris.beams.io/posts/git-commit/). 
  Summarized as follows:
  1. Separate subject from body with a blank line
  2. Limit the subject line to 50 characters
  3. Do not end the subject line with a period
  4. Use the imperative mood in the subject line (add, not adding)
  5. Wrap the body at 72 characters
  6. Use the body to explain what and why vs. how
     This implements a framework for API calls to Airflow. Currently
     all access is done by cli or web ui. Especially in the context
     of the cli this raises security concerns which can be alleviated
     with a secured API call over the wire.

Secondly integration with other systems is a bit harder if you have
to call a cli. For public facing endpoints JSON is used.

As an example the trigger_dag functionality is now made into a
API call.

Backwards compat is retained by switching to a LocalClient.
- Docs and tests to be added
- Kerberos integration works

### CODE REVIEW

**Comment by alexvanboxel** [comment]
I haven't tested it yet but I like the idea. Because I was thinking about external triggers (iso sensors waiting for existence of stuff). I could then integrate with some other GCP stuff by having another container running and translating arbitrary web-hooks into triggers.

**Comment by bolkedebruin** [comment]
@alexvanboxel Thanks. Im looking for feed back on the approach. Im not sure how to remove the code redundancy between auth for the api and the login page (and if we want to at all). Also there was some earlier deliberation on what kind of protocol should be exposed for the internal API (service to service), ie. avro or protobuf.

**Comment by criccomini** [comment]
Awesome to see this getting started. High-level looks good to me.

> Im not sure how to remove the code redundancy between auth for the api and the login page (and if we want to at all).

Does it make sense to think that the UI will eventually make REST calls to these end points?

> Also there was some earlier deliberation on what kind of protocol should be exposed for the internal API (service to service), ie. avro or protobuf.

Yea, I can't find this discussion anymore (probably on Gitter, sigh). Was there any reason not to use JSON? Another crazy idea would be gRPC.

**Comment by criccomini** [comment]
Also, any high-level docs on thins? Specifically, thoughts around security are useful.

**Comment by bolkedebruin** [comment]
Hey @criccomini nice to see you around again ;-). On the authentication things, yes I do think we might move to an UI eventually that just does rest calls. Maybe I was just over complicating things in my head or I was looking at the kerberos stuff that is different from when you login as a user or when you expose it as a service. Need to to think about that a little bit more. We be great to talk about it soon. Input is needed :). 

I wrote a little bit on the implementation but it obviously needs more attention. Maybe you can add to it? https://cwiki.apache.org/confluence/display/AIRFLOW/API+Conventions

**Comment by dmikov** [comment]
Hi, guys. Do you have an ETA on API? I am currently writing a plugin to parse custom definitions into dags, but need more API for front end and was about to hack something myself using sql alchemy connections to DB. But excited that I might not need to do it. 
Do you need any help? I can at least help with tedious Swagger interface definitions, which now becoming de-facto for REST discovery.  [https://github.com/OAI/OpenAPI-Specification](https://github.com/OAI/OpenAPI-Specification) I use, flask-swagger in my plugin it's mostly manual, but at least in place.

**Comment by bolkedebruin** [comment]
@dmikov Well it depends on what you consider API. Fully moving everything over in one commit is not what I am aiming at. I would like to lay down the fundaments and then hopefully others will move things over that they require. 

We probably won't accept new functionality in new PRs that do not use the API framework.

I could definitely use some help. The swagger interface seems smart to integrate. So if you can take that on that would be nice. Also documentation on how to do the API properly needs some love (and thinking) - https://cwiki.apache.org/confluence/display/AIRFLOW/API+Conventions .

**Comment by amcdnl** [comment]
Is there time-frame we can expect this to be merged in?

**Comment by r39132** [comment]
+1

**Comment by r39132** [comment]
@bolkedebruin Let me know your intentions toward getting this PR across the line in the next 5 days.

**Comment by bolkedebruin** [comment]
I will simplify the api by removing the swagger part. This can be added later. Then I will merge.

**Comment by r39132** [comment]
@bolkedebruin Great. Do you have a date in mind for the merge? Should we try to get this into 1.8?

**Comment by r39132** [comment]
@bolkedebruin This has been open for 2.5 months. Any chance you can get this in? Else, I'll close.

**Comment by bolkedebruin** [comment]
@r39132 This isn't done yet, it was also dependent on other PRs that just got in (trusty env). And I don't think it should be closed because it is slow to progress.

**Comment by amcdnl** [comment]
I'd vote no to closing, this is something I think the community needs and @bolkedebruin has done a great job getting this started.

**Comment by bolkedebruin** [comment]
Should build fine now (maybe some license stuff left). Left todo (but can be separate PR):
- Add kerberos tests, can be done now as on Trusty and non docker
- Add swagger definitions: need some help here as I am not familiar with it

File: tests/www/api/experimental/test_endpoints.py [hunk 1]
```diff
@@ -52,10 +52,9 @@ def reset(dag_id=TEST_DAG_ID):
 
 class ApiExperimentalTests(unittest.TestCase):
     def setUp(self):
-        reset()
-        configuration.load_test_config()
-        app = application.create_app()
-        app.config['TESTING'] = True
+        #reset()
```
Comment (ldct) [inline tests/www/api/experimental/test_endpoints.py]: are these supposed to be commented out?

Comment (bolkedebruin) [reply tests/www/api/experimental/test_endpoints.py]: yes, but will remove completely

File: tests/www/api/experimental/test_kerberos_endpoints.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+        # disable mutual authentication for testing
+        client_auth.mutual_authentication = 3
+
+        # case can influence the results
+        client_auth.hostname_override = socket.getfqdn()
+
+        client_auth.handle_response(response)
+
+        print(response.request.headers)
```
Comment (ldct) [inline tests/www/api/experimental/test_kerberos_endpoints.py]: assert something about headers instead of printing them

**Comment by bolkedebruin** [comment]
@alexvanboxel if you have a chance, please review. Some attention to the testing code would be nice as I am forced to reload some stuff in app.py just for testing. Nose keeps some things loaded into mem and I dont know a way around it except for this hack.

**Comment by r39132** [comment]
@bolkedebruin Thanks for working on this and getting this across the line. Strictly speaking, we do require a +1 from another committer before merging new functionality - exceptions being hotfixes/reversions.

**Comment by bolkedebruin** [comment]
@r39132 absolutely right, I thought I had one after the many discussions on this, but that wasn't the case.

**Pull request merged**